### ISSUE TITLE
Disabling direct render causes crash in editor

### ISSUE THREAD

**Issue by sh3rifme** [googlesamples/cardboard-unity#45]
I noticed that when i disable direct render to get bloom working on the main camera, i cannot swap from 2D to VR mode while running in the editor as Unity will crash immediately. It also crashes if direct render is enabled and it's running in VR mode, and then direct render is disabled.

I'm pretty sure i had it working with direct render off before, I've been using the bloom script for a while now, although i only updated to v5.0 this morning.

Works either way on the phone, this appears to be an editor-only issue.

I tried re-importing the entire plugin (delete cardboard and plugins folder, import again) still crashes.

**Comment by smdol** [comment]
Was the bloom effect attached to the parent main camera or to the stereo left and right cameras?

**Comment by smdol** [comment]
Additional question: which version of the Cardboard SDK are you using?

**Comment by sh3rifme** [comment]
Hey, thanks for the fast replies! It's attached to the 'Main Camera' parent component. Ah, i meant to say v0.5 (of cardboard), not v5.0!

**Comment by sh3rifme** [comment]
I think it's worth noting the issue occurs even when there are no post-processing effects attached to the main camera.

**Comment by smdol** [comment]
OK.  I am not able to reproduce this error.  Can you look in the editor log file to see if it manages to put out any error messages before it disappears?

**Comment by sh3rifme** [comment]
Hey sorry it's taken me so long to get back to you, I've taken the log output and put it into [this pastebin.](http://pastebin.com/YKm3ebc4)

**Comment by smdol** [comment]
This looks like an important line:
    RenderTexture.Create failed: format unsupported.

Does this only happen when the bloom effect is being used?  Is it the standard Unity Bloom script, or some other one?  

I guess the key thing to figure out know is that format.  There might be one in the bloom script.  Also, the default format for the Cardboard stereo texture is RGB565, which you can change in the BaseVrDevice.CreateStereoScreen() function, if you want to try that.

**Comment by MiCkSoftware** [comment]
Hi!

I had the same issue some time ago. It seems that  render texture has some
problem with HDR cameras...

Cheers.
MiCk.

**Comment by thesambassador** [comment]
I'm hitting this issue too.  To reproduce, I just downloaded the Cardboard SDK unity package, the Cardboard Demo unity package, and imported them both into a project.  Then I opened up the demo scene, went to the StereoController component on the main camera, and unchecked "Direct Render."  This causes it to crash for me 100% of the time (I have not added any image effects or done anything else to the demo scene).

**Comment by smdol** [comment]
@thesambassador What's your version of Unity?

**Comment by thesambassador** [comment]
Actually, I went ahead and upgraded to the latest (5.1.3f1) and now it doesn't crash, but instead when I start it's throwing a bunch of errors.

RenderTexture.Create failed: format unsupported.
UnityEngine.Camera:Render()
CardboardEye:Render() (at Assets/Cardboard/Scripts/CardboardEye.cs:239)
StereoController:OnPreCull() (at Assets/Cardboard/Scripts/StereoController.cs:330)

Going into CardboardEye.cs, if I change the line
RenderTextureFormat format = stereoScreen ? stereoScreen.format : RenderTextureFormat.RGB565;

and change the RenderTextureFormat to RenderTextureFormat.Default, it "works", but doesn't look right (the sky covers other objects in the scene, the sky is black, the cardboard curved overlay is rendered funny).

Any ideas?

**Comment by smdol** [comment]
Can you post a screenshot?  That's pretty wacky-sounding, and I don't think I've seen it before.

**Comment by thesambassador** [comment]
Here's what it looks like when I launch it just by disabling DirectRender:

![image](https://cloud.githubusercontent.com/assets/119074/9569791/5f5e11bc-4f2a-11e5-954d-e03f8e33715a.png)

Here's what it looks like when I change the RenderTextureFormat to Default:
![image](https://cloud.githubusercontent.com/assets/119074/9569794/903a8658-4f2a-11e5-8242-1b675614da3f.png)

**Comment by smdol** [comment]
As an experiment, can you change your Build Platform to Android?

**Comment by thesambassador** [comment]
No change, same as before.

**Comment by smdol** [comment]
OK, experiment #2:  Direct Render off, Use RenderTextureFormat.Default, and in turn off DistortionCorrection on the CardboardMain object.  (This just turns off the Image Effect that Unity is using in the Editor for distortion -- one less step in the pipeline.)

**Comment by thesambassador** [comment]
Still seems the same... works, but the background is black and the particles out of the cube seem to not be rendering correctly.

**Comment by smdol** [comment]
I'm looking at your screenshot with the black sky.  Something odd -- there's a faint bluish outline around the edge of the sky, where the vignette is supposed to be.  It's like the blue sky is being rendered behind the black sky (which is also obscuring the particles).

**Comment by thesambassador** [comment]
Yeah, it's really odd.  Not sure what is going on there.  I haven't changed anything in the demo scene, and it renders fine with the direct render.

**Comment by smdol** [comment]
OK, this is definitely strange.  It seems to be specific to Windows (I have a Mac and don't experience any issue).  So I'm shooting in the dark a bit, but let's see what we can do to characterize the problem a bit more.

Experiment #3: In the Cardboard.cs function AddDummyCamera(), change the color from black to something else.  Does the mysterious black sky follow suit, or stay black?

Experiment #4: Comment out the use of AddDummyCamera() altogether.  Does the problem still occur?

Experiment #5:  In Player Settings, in the tab for Windows, clear the "Use DirectX 11" checkbox.  This makes Unity use DirectX 9.  Still get black sky?  If no, can you go back to RGB565?

Experiment #6:  Start Unity with --force-opengl (you'll have to edit the shortcut properties).  You won't see <DX11> in the title bar now.  Same question as for #5.

In #3 and #4 I'm looking at what's going on in the depth buffer.  In #5 and #6 it's about the graphics API the editor is using.

**Comment by thesambassador** [comment]
Experiment 3 - Yes, I changed it to Color.green and the background changed to green!

Experiment 4 - Commenting out the AddDummyCamera seems to make the background pretty much gone, but adds a weird trail effect too:
![image](https://cloud.githubusercontent.com/assets/119074/9570190/fbc94d98-4f37-11e5-8d5a-f61b00e240f5.png)

Experiment 5 - I can't seem to set my DirectX settings in 5.1.3.  According to the "what's new in Unity 5.1", they mention replacing the "Use DX11" checkbox with another setting, but I can't find what they are talking about.
https://unity3d.com/unity/whats-new/unity-5.1

Experiment 6 - Even with --force-opengl it seems like I have the same issue (black sky with default, error with RGB565)

**Comment by thesambassador** [comment]
Ah wait!  I figured out how to force DX 9.  Setting it to DX 9 seems to be working with RGB565!

So I have a workaround, but it looks like DX 11 is the problem!

**Comment by smdol** [comment]
I figured out the weird "Black Sky" effect you were seeing when using RenderTextureFormat.Default:  If the render texture format has an alpha channel, then the non-direct-render path will blend the temporary render texture onto the screen.  The default blue sky color on Main Camera has a nearly transparent alpha value, so it is getting blended away.  I changed the alpha to 255 in the Main Camera's sky color, and the black sky effect is gone.
